OpenSim/Common/DataTable.h

```diff
@@ -0,0 +1,1142 @@
+
+  /// Construct DataTable using an iterator(satisfying requirement of an 
+  /// input_iterator) which produces one entry at a time. The entries of 
+  /// DataTable_ are copy initialized using the values produced by the 
+  /// iterator. For example, specifying ROWWISE for parameter dir and 10 for 
+  /// parameter ndir will populate the DataTable one row at a time with each 
+  /// row's length taken to be 10 elements.
+  ///
+  /// \param first Beginning of range covered by the iterator.
+  /// \param last End of the range covered by the iterator.
+  /// \param ndir Extent of the dimention specified by parameter dir. 
+  /// \param dir Dimension to populate the DataTable. Populate row-wise or 
+  ///              column wise. See OpenSim::InputItDir for possible values.
+  /// \param allow_missing Allow for missing values. When set to false, this
+  ///                      function will throw if the input iterator fills up 
+  ///                      the last row/col partially. When set to true, 
+  ///                      missing elements are set to SimTK::NaN.
+  ///
+  /// \throws ZeroElements When input-iterator does not produce any elements.
+  ///                      That is first == last.
+  template<typename InputIt>
+  DataTable_(InputIt first,
+             typename std::enable_if<!std::is_integral<InputIt>::value,
+                                     InputIt>::type last,
+             const size_t ndir,
+             const InputItDir dir = ROWWISE,
+             const bool allow_missing = false) :
+    data{static_cast<int>(dir == ROWWISE ? 1    : ndir), 
+         static_cast<int>(dir == ROWWISE ? ndir :    1)},
+    metadata{},
+    col_ind{} {
+    if(!(first != last))
+      throw ZeroElements{"Input iterators produce zero elements."};
+    if(ndir == 0)
+      throw InvalidEntry{"Input argument 'ndir' cannot be zero."};
+
+    int row{0};
+    int col{0};
+    while(first != last) {
+      data.set(row, col, *first);
+      ++first;
+      if(dir == ROWWISE) {
+        ++col;
+        if(col == static_cast<int>(ndir)  && first != last) {
+          col = 0;
+          ++row;
+          data.resizeKeep(data.nrow() + 1, data.ncol());
+        }
+      } else {
+        ++row;
+        if(row == static_cast<int>(ndir) && first != last) {
+          row = 0;
+          ++col;
+          data.resizeKeep(data.nrow(), data.ncol() + 1);
+        }
+      }
+    }
+    if(!allow_missing) {
+      if((dir == ROWWISE && col != data.ncol()) ||
+         (dir == COLWISE && row != data.nrow()))
+        throw InvalidEntry{"Input iterator did not produce enough elements to "
+                           "fill the last row."};
+    }
+  }
+
+  /// Copy constructor. Perform a deep copy of the DataTable_.
+  DataTable_(const DataTable_& dt) :
+    data{dt.data}, metadata{dt.metadata}, col_ind{dt.col_ind} {
+      // Deep copy of metadata.
+      for(auto& elem : metadata)
+        elem.second = elem.second->clone();
```

Comment:
This could have used a default copy constructor but for the metadata cloning. If you use `ClonePtr<AbstractValue>` as the type of `second` then you can remove this code altogether, which is a very good thing. Using default copy constructors is the only way to make sure they continue to work when the data members get changed (probably by some other programmer!).